design-proposals/klip-55-tablespaces.md

```diff
@@ -1,3 +1,210 @@
+a subset of those objects have a specific usage pattern. If you could tune for
+that usage pattern on a granular level, ksqlDB could treat those objects more efficiently.
+
+Tablespaces are a construct for defining groups of configuration that can be applied
+to one or more database objects. They exist in
+[Postgres](https://www.postgresql.org/docs/10/manage-ag-tablespaces.html),
+[MySQL](https://dev.mysql.com/doc/refman/8.0/en/general-tablespaces.html),
+[Oracle](https://docs.oracle.com/cd/A57673_01/DOC/server/doc/SCN73/ch4.htm), and so on.
+In the context on ksqlDB, a tablespace would allow you to override default server
+configuration for one or more objects.
+
+This is particularly important given the impact on how ksqlDB's runtime is being rearchitected.
+When multiple persistent queries begin to share a runtime, it will become more challenging to
+vary configuration parameters across them. Tablespaces give you a logical boundary to hint to
+ksqlDB where it must respect varying parameters.
+
+Tablespaces don't enforce namespacing or scoping: two objects in different tablespaces
+can be used with one another.
+
+## What is in scope
+
+- The grammar to support tablespace commands
+- The behavior of those commands
+- Backward compatibility
+
+## What is not in scope
+
+- How it's implemented
+
+## Value/Return
+
+- Improved program performance
+- Isolation of functional parameters to a subset of objects
+
+## Public APIs
+
+**Using the default tablespace**
+
+Creating an object works like normal—it's placed into a default tablespace so
+that you never need to use this feature unless you want to.
+
+```sql
+CREATE stream object WITH (
+    ...
+);
+```
+
+The default tablespace is `ksql_default`.
+
+**Creating a tablespace**
+
+You can supply one or more key/value parameters. These parameters are server-level
+configuration parameters that you want to group together.
+
+```sql
+CREATE TABLESPACE tablespace_name WITH (
+    property_name = expression [,
+    ...
+    ]
+);
+```
+
+By default, no objects are placed within a new tablespace.
+
+- TODO: what happens if `tablespace_name` already exists?
+
+**Adding a newly created object to a tablespace**
+
+Explicitly define what tablespace a stream or table should reside in.
+
+```sql
+CREATE [STREAM|TABLE] obj WITH (
+    'tablespace' = 'tablespace_name',
+    ...
+);
+```
+
+If `tablespace_name` doesn't exist, `obj`'s creation is rejected with a clear error message why.
+
+**Seeing what tablespace an object is in**
+
+```sql
+DESCRIBE obj EXTENDED;
+```
+
+Output fragment:
+
+```
+Name                 : KSQL_PROCESSING_LOG
+Type                 : STREAM
+Timestamp field      : Not set - using <ROWTIME>
+Tablespace           : tablespace_name
+```
+
+**Setting a default tablespace**
+
+You can override the default tablespace so you don't
+need to supply it to every proceeding object that gets created.
+
+```sql
+SET 'ksql.default.tablespace' = 'tablespace_name';
```

Comment:
i think a better analogy here than setting the value of a "session variable" like that is to take inspiration from the rdbms world: `USE space-name;`